src/datatype/functions_vecf32.rs

```diff
@@ -0,0 +1,125 @@
+use crate::datatype::memory_vecf32::{Vecf32Input, Vecf32Output};
+use crate::error::*;
+use base::scalar::*;
+use base::vector::*;
+
+/// accumulate intermediate state for vector average
+#[pgrx::pg_extern(immutable, strict, parallel_safe)]
+fn _vectors_vector_accum(
+    mut state: pgrx::composite_type!('static, "vectors.vector_accum_state"),
+    value: Vecf32Input<'_>,
+) -> pgrx::composite_type!('static, "vectors.vector_accum_state") {
+    let count = state
+        .get_by_name::<i64>("count")
+        .unwrap()
+        .unwrap_or_default();
+    if count == 0 {
+        // state is empty
+        let mut result =
+            pgrx::heap_tuple::PgHeapTuple::new_composite_type("vectors.vector_accum_state")
+                .unwrap();
+        let sum = value.iter().map(|x| x.0 as f64).collect::<Vec<_>>();
+        result.set_by_name("count", count + 1).unwrap();
+        result.set_by_name("sum", sum).unwrap();
+        result
+    } else {
+        let sum = state
+            .get_by_name::<pgrx::Array<f64>>("sum")
+            .unwrap()
+            .unwrap();
+        check_matched_dims(sum.len(), value.dims());
+        // TODO: pgrx::Array<T> don't support mutable operations currently, we can reuse the state once it's supported.
```

Comment:
I double check this solution. It seem that the memory management of `AggState` is well-designed. For `transfn` like `sfunc( internal-state, next-data-values ) ---> next-internal-state` , return `internal-state` after modification as result is just okay.
See https://github.com/postgres/postgres/blob/52b49b796cc7fd976f4da6aa49c9679ecdae8bd5/src/backend/executor/nodeAgg.c#L81-L100 and https://doxygen.postgresql.org/int8_8c.html#aff6c10f1da41c3aca8253c3e1d6f1378 for more details.
I am going to change the `composite_type` to a struct. And modify the state in-place.